examples/toy_webrtc.rs

```diff
@@ -0,0 +1,188 @@
+    dbg!(deser);
+    todo!();
+    */
+
+    match args
+        .next()
+        .expect("Role argument is required: server or client")
+        .as_ref()
+    {
+        "server" => run_server(),
+        "client" => run_client(),
+        _ => panic!("Argument must be server or client"),
+    }
+}
+
+/// The server is just an echo client, with some chaos introduced
+fn run_server() -> std::io::Result<()> {
+    let socket = UdpSocket::bind(SERVER_ADDR)?;
+    println!("Server listening at {}", socket.local_addr()?);
+
+    let mut buf = [0; DATA_SIZE];
+    loop {
+        let (amt, src) = socket.recv_from(&mut buf)?;
+        let buf = &mut buf[..amt];
+
+        // introduce chaos
+        /*
+        use rand::Rng;
+        let mut rng = rand::thread_rng();
+        match rng.gen_range(0..100) {
+            0 => continue, // 1% packet loss
+            1 => std::thread::sleep(std::time::Duration::from_millis(1)), // 1% of packets has 1ms delay
+            _ => (),
+        };
+        */
+
+        socket.send_to(buf, &src)?;
+    }
+}
+
+// cram 128 floats with a bit of overhead into 508 bytes
+fn serialize(audio_buf: &AudioBuffer, byte_buf: &mut [u8]) -> usize {
+    // convert f32 to i16, get big endian bytes, write to byte_buf[0..256]
+    let n = audio_buf
+        .get_channel_data(0)
+        .iter()
+        .map(|f| (f * i16::MAX as f32) as i16)
+        .map(|i| i.to_be_bytes())
+        .flat_map(IntoIterator::into_iter)
+        .zip(byte_buf.iter_mut())
+        .map(|(i, o)| *o = i)
+        .count();
+
+    assert!(n < MAX_UDP_SIZE);
+
+    n
+}
+
+fn deserialize(byte_buf: &[u8], sample_rate: SampleRate) -> AudioBuffer {
+    let samples: Vec<f32> = byte_buf
+        .chunks_exact(2)
+        .take(128)
+        .map(|bs| i16::from_be_bytes(bs.try_into().unwrap()))
+        .map(|i| i as f32 / i16::MAX as f32)
+        .collect();
+    AudioBuffer::from(vec![samples], sample_rate)
+}
+
+struct SocketStream {
+    socket: &'static UdpSocket,
+    sample_rate: SampleRate,
+    byte_buf: Vec<u8>,
+}
+
+impl Iterator for SocketStream {
+    type Item = Result<AudioBuffer, Box<dyn Error + Send + Sync>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let bytes = match self.socket.recv(&mut self.byte_buf) {
+            Ok(received) => &self.byte_buf[..received],
+            Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+                // construct empty buffer
+                let options = AudioBufferOptions {
+                    number_of_channels: 1,
+                    length: RENDER_QUANTUM_SIZE,
+                    sample_rate: self.sample_rate,
+                };
+                return Some(Ok(AudioBuffer::new(options)));
+            }
+            Err(e) => panic!("client recv IO error: {}", e),
+        };
+        Some(Ok(deserialize(bytes, self.sample_rate)))
+    }
+}
+
+/// The client records their audio, ships it to the server and plays the received samples
```

Comment:
"The client 1. records the audio and ships it to the server, and 2. plays back the samples received from the server"